### PR TITLE
Added Corporea Autocompletion system

### PR DESCRIPTION
**Pull request by soundlogic2236**
Lag free with 120 mods and facades!

### CODE REVIEW

File: src/main/java/vazkii/botania/api/corporea/CorporeaHelper.java [hunk 1]
```diff
@@ -314,4 +315,21 @@ public static boolean arrayHas(int[] arr, int val) {
...
+	public static void registerAutoCompleteController(ICorporeaAutoCompleteController controller) {
+		autoCompleteControllers.add(controller);
+	}
+
+	/**
+	 * Returns if the auto complete helper should run
+	 */
+	public static boolean shouldAutoComplete() {
+		for(ICorporeaAutoCompleteController controller : autoCompleteControllers)
+			if(controller.shouldAutoComplete())
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/api/corporea/CorporeaHelper.java]: Just return this.

File: src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java [hunk 2]
```diff
@@ -0,0 +1,195 @@
...
+import cpw.mods.fml.relauncher.ReflectionHelper;
+
+public class CorporeaAutoCompleteHandler {
+
+	boolean isAutoCompleted = false;
+	String originalString = "";
+	List<CompletionData> completions = new ArrayList<CompletionData>();
+	int position;
+
+	private static class CompletionData implements Comparable<CompletionData> {
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: Move this to the bottom.

File: src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+			});
+
+	private boolean tabLastTick = false;
+
+	public static void updateItemList() {
+		itemNames.clear();
+		Iterator<Item> iterator = Item.itemRegistry.iterator();
+		ArrayList<ItemStack> curList = new ArrayList<ItemStack>();
+
+		while (iterator.hasNext()) {
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: The spaces sound, the spaces.

File: src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+			return;
+		GuiScreen screen = Minecraft.getMinecraft().currentScreen;
+		if (!(screen instanceof GuiChat)) {
+			isAutoCompleted = false;
+			return;
+		}
+		GuiChat chat = (GuiChat) screen;
+		if (isAutoCompleted) {
+			boolean valid = ReflectionHelper.getPrivateValue(GuiChat.class, chat, LibObfuscation.COMPLETE_FLAG);
+			if (!valid) {
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: Needless brackets.

File: src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+		if (this.position >= completions.size())
+			this.position -= completions.size();
+		CompletionData data = completions.get(position);
+		String str = originalString.substring(0, originalString.length() - data.prefixLength) + data.string;
+		inputField.setText(str);
+	}
+
+	private void buildAutoCompletes(GuiTextField inputField, GuiChat chat) {
+		String leftOfCursor;
+		if (inputField.getCursorPosition() == 0) {
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: Bunch of needless brackets.

File: src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+		this.position = -1;
+		ReflectionHelper.setPrivateValue(GuiChat.class, chat, true, LibObfuscation.COMPLETE_FLAG);
+        StringBuilder stringbuilder = new StringBuilder();
+        CompletionData data;
+        for (Iterator<CompletionData> iterator = completions.iterator(); iterator.hasNext(); stringbuilder.append(data.string))
+        {
+            data = iterator.next();
+
+            if (stringbuilder.length() > 0)
+            {
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: More.

File: src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+                stringbuilder.append(", ");
+            }
+        }
+        
+        Minecraft.getMinecraft().ingameGUI.getChatGUI().printChatMessageWithOptionalDeletion(new ChatComponentText(stringbuilder.toString()), 1);
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: Bad Indenting.

File: src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+	private List<CompletionData> getNamesStartingWith(String prefix) {
+		ArrayList<CompletionData> result = new ArrayList<CompletionData>();
+		int length = prefix.length();
+		SortedSet<String> after = itemNames.tailSet(prefix);
+		for (String str : after) {
+			if (str.toLowerCase().startsWith(prefix))
+				result.add(new CompletionData(str, length));
+			else
+				return result;
```
Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: Same line as the else.

File: src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java (hunk 2 above)
Comment (soundlogic2236) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: The class?

File: src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java (hunk 3 above)
Comment (soundlogic2236) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: ...where? I thought I got all the spacing around =s and such.

File: src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java (hunk 2 above)
Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: I usually have the internal classes in the bottom, yes.

File: src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java (hunk 3 above)
Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: After the while. All the other statements before this don't have a space.

File: src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java (hunk 4 above)
Comment (soundlogic2236) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: In the instanceof bit?

Comment (Vazkii) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: This if only have one line, you don't need brackets.

Comment (soundlogic2236) [inline src/main/java/vazkii/botania/client/core/handler/CorporeaAutoCompleteHandler.java]: Oh! The !valid bit. github broke that up. I shall fix

**Pull request closed without merging**